File: python/social_tinder/analytics.py
"""Graph algorithms used to enrich the keyword network.

All operate on a simple weighted undirected graph expressed as:
  - node_ids: list[str]
  - adj: dict[str, list[tuple[str, float]]]   (symmetric; weight = edge strength)

Implemented here (pure-stdlib, deterministic):
  - log_likelihood_ratio : Dunning's G² for 2x2 contingency  -> edge significance
  - pagerank             : weighted PageRank                  -> keyword influence
  - betweenness          : Brandes' algorithm (unweighted)    -> bridge score
  - louvain              : modularity maximization            -> communities
"""

from __future__ import annotations

import math
from collections import deque


# --- Edge significance ------------------------------------------------------

def _g2_cell(observed: float, expected: float) -> float:
    if observed <= 0 or expected <= 0:
        return 0.0
    return observed * math.log(observed / expected)


def log_likelihood_ratio(k11: int, count_a: int, count_b: int, n: int) -> float:
    """Dunning's log-likelihood ratio (G²) for the 2x2 table of A vs B presence.

    k11 = mentions with both, count_a/count_b = marginals, n = total mentions.
    Larger => the co-occurrence is less likely to be a coincidence.
    """
    if n <= 0:
        return 0.0
    k12 = count_a - k11        # A without B
    k21 = count_b - k11        # B without A
    k22 = n - count_a - count_b + k11  # neither
    e11 = count_a * count_b / n
    e12 = count_a * (n - count_b) / n
    e21 = (n - count_a) * count_b / n
    e22 = (n - count_a) * (n - count_b) / n
    g2 = 2.0 * (
        _g2_cell(k11, e11)
        + _g2_cell(k12, e12)
        + _g2_cell(k21, e21)
        + _g2_cell(k22, e22)
    )
    return max(0.0, g2)


# --- PageRank ---------------------------------------------------------------

def pagerank(
    node_ids: list[str],
    adj: dict[str, list[tuple[str, float]]],
    damping: float = 0.85,
    max_iter: int = 200,
    tol: float = 1e-9,
) -> dict[str, float]:
    """Weighted PageRank. Returns scores summing to ~1 (uniform if no edges)."""
    n = len(node_ids)
    if n == 0:
        return {}
    out_weight = {i: sum(w for _, w in adj.get(i, [])) for i in node_ids}
    pr = {i: 1.0 / n for i in node_ids}
    base = (1.0 - damping) / n

    for _ in range(max_iter):
        dangling = sum(pr[i] for i in node_ids if out_weight[i] == 0.0)
        new = {i: base + damping * dangling / n for i in node_ids}
        for i in node_ids:
            wi = out_weight[i]
            if wi == 0.0:
                continue
            share = damping * pr[i] / wi
            for j, w in adj[i]:
                new[j] += share * w
        diff = sum(abs(new[i] - pr[i]) for i in node_ids)
        pr = new
        if diff < tol:
            break
    return pr


# --- Betweenness centrality (Brandes, unweighted) ---------------------------

def betweenness(
    node_ids: list[str],
    adj: dict[str, list[tuple[str, float]]],
) -> dict[str, float]:
    """Normalized betweenness in [0, 1]. Identifies 'bridge' keywords that sit
    on many shortest paths between communities. Treats edges as unweighted."""
    cb = {v: 0.0 for v in node_ids}
    neighbors = {v: [j for j, _ in adj.get(v, [])] for v in node_ids}

    for s in node_ids:
        stack: list[str] = []
        pred: dict[str, list[str]] = {v: [] for v in node_ids}
        sigma = {v: 0.0 for v in node_ids}
        dist = {v: -1 for v in node_ids}
        sigma[s] = 1.0
        dist[s] = 0
        q: deque[str] = deque([s])
        while q:
            v = q.popleft()
            stack.append(v)
            for w in neighbors[v]:
                if dist[w] < 0:
                    dist[w] = dist[v] + 1
                    q.append(w)
                if dist[w] == dist[v] + 1:
                    sigma[w] += sigma[v]
                    pred[w].append(v)
        delta = {v: 0.0 for v in node_ids}
        while stack:
            w = stack.pop()
            for v in pred[w]:
                if sigma[w] > 0:
                    delta[v] += (sigma[v] / sigma[w]) * (1.0 + delta[w])
            if w != s:
                cb[w] += delta[w]

    # Undirected: each path counted twice. Normalize by max possible pairs.
    n = len(node_ids)
    scale = (n - 1) * (n - 2) / 2.0 if n > 2 else 0.0
    for v in node_ids:
        cb[v] /= 2.0
        cb[v] = cb[v] / scale if scale > 0 else 0.0
    return cb


# --- Louvain community detection -------------------------------------------

def louvain(
    node_ids: list[str],
    adj: dict[str, list[tuple[str, float]]],
    resolution: float = 1.0,
    max_passes: int = 20,
) -> dict[str, int]:
    """Deterministic Louvain modularity maximization. Returns node -> community
    index (small ints). Falls back to singletons when there are no edges."""
    if not node_ids:
        return {}

    # Aggregate self-loops as we coarsen; start with the raw graph.
    # graph: list of node keys; weights between super-nodes; self-loop weights.
    keys = list(node_ids)
    # adjacency as dict[key][key] = weight (undirected, stored once each dir)
    g: dict[str, dict[str, float]] = {k: {} for k in keys}
    for i in keys:
        for j, w in adj.get(i, []):
            g[i][j] = g[i].get(j, 0.0) + w
    self_loops: dict[str, float] = {k: 0.0 for k in keys}

    # Track the mapping from original nodes to current super-nodes.
    membership = {i: i for i in node_ids}

    while True:
        comm, improved = _louvain_one_level(keys, g, self_loops, resolution)
        # Re-index communities of this level to contiguous ids.
        # Update original-node membership.
        for orig, sup in membership.items():
            membership[orig] = comm[sup]
        if not improved:
            break
        # Build aggregated graph keyed by community id.
        new_keys = sorted(set(comm.values()))
        ng: dict[str, dict[str, float]] = {k: {} for k in new_keys}
        nsl: dict[str, float] = {k: 0.0 for k in new_keys}
        for i in keys:
            ci = comm[i]
            nsl[ci] += self_loops[i]
            for j, w in g[i].items():
                cj = comm[j]
                if ci == cj:
                    nsl[ci] += w  # counted once per direction -> internal*2 total
                else:
                    ng[ci][cj] = ng[ci].get(cj, 0.0) + w
        keys, g, self_loops = new_keys, ng, nsl
        if len(new_keys) == 1:
            break

    # Re-index final communities to 0..k-1 in a stable order.
    order_seen: dict[int, int] = {}
    nxt = 0
    result: dict[str, int] = {}
    for i in node_ids:  # original order is deterministic
        c = membership[i]
        if c not in order_seen:
            order_seen[c] = nxt
            nxt += 1
        result[i] = order_seen[c]
    return result


def _louvain_one_level(
    keys: list[str],
    g: dict[str, dict[str, float]],
    self_loops: dict[str, float],
    resolution: float,
) -> tuple[dict[str, int], bool]:
    """One pass of local moving. Returns (node->community-int, improved?)."""
    # Degree (k_i) includes self-loop weight twice.
    k = {i: self_loops[i] * 2 + sum(g[i].values()) for i in keys}
    m = sum(k.values()) / 2.0  # total edge weight
    if m == 0:
        return {i: idx for idx, i in enumerate(keys)}, False

    community = {i: i for i in keys}
    tot = {i: k[i] for i in keys}  # sum of degrees in each community

    improved_any = False
    moved = True
    passes = 0
    while moved and passes < 50:
        moved = False
        passes += 1
        for i in keys:
            ci = community[i]
            # weights from i to each neighboring community
            nbr_w: dict[str, float] = {}
            for j, w in g[i].items():
                if j == i:
                    continue
                nbr_w[community[j]] = nbr_w.get(community[j], 0.0) + w
            # Remove i from its community.
            tot[ci] -= k[i]
            ki_in_ci = nbr_w.get(ci, 0.0)
            # Find best community (deterministic: best gain, tie -> smallest id).
            best_c, best_gain = ci, ki_in_ci - resolution * tot[ci] * k[i] / (2.0 * m)
            for c, w_in in nbr_w.items():
                gain = w_in - resolution * tot[c] * k[i] / (2.0 * m)
                if gain > best_gain or (gain == best_gain and str(c) < str(best_c)):
                    best_c, best_gain = c, gain
            tot[best_c] += k[i]
            community[i] = best_c
            if best_c != ci:
                moved = True
                improved_any = True

    # Re-index this level's communities to contiguous ints.
    remap: dict[str, int] = {}
    nxt = 0
    out: dict[str, int] = {}
    for i in keys:
        c = community[i]
        if c not in remap:
            remap[c] = nxt
            nxt += 1
        out[i] = remap[c]
    return out, improved_any


def modularity(
    node_ids: list[str],
    adj: dict[str, list[tuple[str, float]]],
    community: dict[str, int],
    resolution: float = 1.0,
) -> float:
    """Newman modularity Q of a partition — used for tests/validation."""
    m2 = sum(w for i in node_ids for _, w in adj.get(i, []))  # = 2m
    if m2 == 0:
        return 0.0
    deg = {i: sum(w for _, w in adj.get(i, [])) for i in node_ids}
    q = 0.0
    for i in node_ids:
        for j, w in adj.get(i, []):
            if community[i] == community[j]:
                q += w - resolution * deg[i] * deg[j] / m2
    return q / m2

File: src/lib/network/analytics.ts
// Graph algorithms that enrich the keyword network. Ports of the Python
// social_tinder/analytics.py — kept in lockstep so both editions agree.
//
// All operate on a weighted undirected graph:
//   nodeIds: string[]
//   adj: Map<string, [string, number][]>   (symmetric; weight = edge strength)

// --- Edge significance ------------------------------------------------------

function g2Cell(observed: number, expected: number): number {
  if (observed <= 0 || expected <= 0) return 0;
  return observed * Math.log(observed / expected);
}

/**
 * Dunning's log-likelihood ratio (G²) for the 2x2 table of A vs B presence.
 * Larger => the co-occurrence is less likely to be coincidental.
 */
export function logLikelihoodRatio(
  k11: number,
  countA: number,
  countB: number,
  n: number,
): number {
  if (n <= 0) return 0;
  const k12 = countA - k11; // A without B
  const k21 = countB - k11; // B without A
  const k22 = n - countA - countB + k11; // neither
  const e11 = (countA * countB) / n;
  const e12 = (countA * (n - countB)) / n;
  const e21 = ((n - countA) * countB) / n;
  const e22 = ((n - countA) * (n - countB)) / n;
  const g2 =
    2 * (g2Cell(k11, e11) + g2Cell(k12, e12) + g2Cell(k21, e21) + g2Cell(k22, e22));
  return Math.max(0, g2);
}

// --- PageRank ---------------------------------------------------------------

/** Weighted PageRank. Returns scores summing to ~1 (uniform if no edges). */
export function pagerank(
  nodeIds: string[],
  adj: Map<string, [string, number][]>,
  damping = 0.85,
  maxIter = 200,
  tol = 1e-9,
): Map<string, number> {
  const n = nodeIds.length;
  const pr = new Map<string, number>();
  if (n === 0) return pr;

  const outWeight = new Map<string, number>();
  for (const i of nodeIds) {
    outWeight.set(i, (adj.get(i) ?? []).reduce((s, [, w]) => s + w, 0));
    pr.set(i, 1 / n);
  }
  const base = (1 - damping) / n;

  for (let iter = 0; iter < maxIter; iter++) {
    let dangling = 0;
    for (const i of nodeIds) if (outWeight.get(i) === 0) dangling += pr.get(i)!;
    const next = new Map<string, number>();
    for (const i of nodeIds) next.set(i, base + (damping * dangling) / n);
    for (const i of nodeIds) {
      const wi = outWeight.get(i)!;
      if (wi === 0) continue;
      const share = (damping * pr.get(i)!) / wi;
      for (const [j, w] of adj.get(i)!) next.set(j, next.get(j)! + share * w);
    }
    let diff = 0;
    for (const i of nodeIds) diff += Math.abs(next.get(i)! - pr.get(i)!);
    for (const i of nodeIds) pr.set(i, next.get(i)!);
    if (diff < tol) break;
  }
  return pr;
}

// --- Betweenness centrality (Brandes, unweighted) ---------------------------

/**
 * Normalized betweenness in [0, 1]. Identifies "bridge" keywords on many
 * shortest paths between communities. Treats edges as unweighted.
 */
export function betweenness(
  nodeIds: string[],
  adj: Map<string, [string, number][]>,
): Map<string, number> {
  const cb = new Map<string, number>(nodeIds.map((v) => [v, 0]));
  const neighbors = new Map<string, string[]>(
    nodeIds.map((v) => [v, (adj.get(v) ?? []).map(([j]) => j)]),
  );

  for (const s of nodeIds) {
    const stack: string[] = [];
    const pred = new Map<string, string[]>(nodeIds.map((v) => [v, []]));
    const sigma = new Map<string, number>(nodeIds.map((v) => [v, 0]));
    const dist = new Map<string, number>(nodeIds.map((v) => [v, -1]));
    sigma.set(s, 1);
    dist.set(s, 0);
    const queue: string[] = [s];
    let head = 0;
    while (head < queue.length) {
      const v = queue[head++];
      stack.push(v);
      for (const w of neighbors.get(v)!) {
        if (dist.get(w)! < 0) {
          dist.set(w, dist.get(v)! + 1);
          queue.push(w);
        }
        if (dist.get(w) === dist.get(v)! + 1) {
          sigma.set(w, sigma.get(w)! + sigma.get(v)!);
          pred.get(w)!.push(v);
        }
      }
    }
    const delta = new Map<string, number>(nodeIds.map((v) => [v, 0]));
    while (stack.length) {
      const w = stack.pop()!;
      for (const v of pred.get(w)!) {
        if (sigma.get(w)! > 0) {
          delta.set(v, delta.get(v)! + (sigma.get(v)! / sigma.get(w)!) * (1 + delta.get(w)!));
        }
      }
      if (w !== s) cb.set(w, cb.get(w)! + delta.get(w)!);
    }
  }

  const n = nodeIds.length;
  const scale = n > 2 ? ((n - 1) * (n - 2)) / 2 : 0;
  for (const v of nodeIds) {
    const halved = cb.get(v)! / 2; // undirected: each path counted twice
    cb.set(v, scale > 0 ? halved / scale : 0);
  }
  return cb;
}

// --- Louvain community detection -------------------------------------------

type WeightMap = Map<string, Map<string, number>>;

/**
 * Deterministic Louvain modularity maximization. Returns node -> community
 * index (small ints).
 */
export function louvain(
  nodeIds: string[],
  adj: Map<string, [string, number][]>,
  resolution = 1,
): Map<string, number> {
  const result = new Map<string, number>();
  if (nodeIds.length === 0) return result;

  let keys = [...nodeIds];
  let g: WeightMap = new Map(keys.map((k) => [k, new Map<string, number>()]));
  for (const i of keys) {
    for (const [j, w] of adj.get(i) ?? []) {
      g.get(i)!.set(j, (g.get(i)!.get(j) ?? 0) + w);
    }
  }
  let selfLoops = new Map<string, number>(keys.map((k) => [k, 0]));
  const membership = new Map<string, string>(nodeIds.map((i) => [i, i]));

  for (;;) {
    const { comm, improved } = louvainOneLevel(keys, g, selfLoops, resolution);
    for (const [orig, sup] of membership) membership.set(orig, String(comm.get(sup)!));
    if (!improved) break;

    const newKeys = [...new Set([...comm.values()])].sort((a, b) => a - b).map(String);
    const ng: WeightMap = new Map(newKeys.map((k) => [k, new Map<string, number>()]));
    const nsl = new Map<string, number>(newKeys.map((k) => [k, 0]));
    for (const i of keys) {
      const ci = String(comm.get(i)!);
      nsl.set(ci, nsl.get(ci)! + selfLoops.get(i)!);
      for (const [j, w] of g.get(i)!) {
        const cj = String(comm.get(j)!);
        if (ci === cj) nsl.set(ci, nsl.get(ci)! + w);
        else ng.get(ci)!.set(cj, (ng.get(ci)!.get(cj) ?? 0) + w);
      }
    }
    keys = newKeys;
    g = ng;
    selfLoops = nsl;
    if (newKeys.length === 1) break;
  }

  // Re-index final communities to 0..k-1 in original-node order.
  const remap = new Map<string, number>();
  let nextId = 0;
  for (const i of nodeIds) {
    const c = membership.get(i)!;
    if (!remap.has(c)) remap.set(c, nextId++);
    result.set(i, remap.get(c)!);
  }
  return result;
}

function louvainOneLevel(
  keys: string[],
  g: WeightMap,
  selfLoops: Map<string, number>,
  resolution: number,
): { comm: Map<string, number>; improved: boolean } {
  const k = new Map<string, number>();
  for (const i of keys) {
    let deg = selfLoops.get(i)! * 2;
    for (const [, w] of g.get(i)!) deg += w;
    k.set(i, deg);
  }
  const m = [...k.values()].reduce((s, v) => s + v, 0) / 2;
  if (m === 0) {
    return { comm: new Map(keys.map((i, idx) => [i, idx])), improved: false };
  }

  const community = new Map<string, string>(keys.map((i) => [i, i]));
  const tot = new Map<string, number>(keys.map((i) => [i, k.get(i)!]));

  let improvedAny = false;
  let moved = true;
  let passes = 0;
  while (moved && passes < 50) {
    moved = false;
    passes++;
    for (const i of keys) {
      const ci = community.get(i)!;
      const nbrW = new Map<string, number>();
      for (const [j, w] of g.get(i)!) {
        if (j === i) continue;
        const cj = community.get(j)!;
        nbrW.set(cj, (nbrW.get(cj) ?? 0) + w);
      }
      tot.set(ci, tot.get(ci)! - k.get(i)!);
      let bestC = ci;
      let bestGain = (nbrW.get(ci) ?? 0) - (resolution * tot.get(ci)! * k.get(i)!) / (2 * m);
      for (const [c, wIn] of nbrW) {
        const gain = wIn - (resolution * tot.get(c)! * k.get(i)!) / (2 * m);
        if (gain > bestGain || (gain === bestGain && c < bestC)) {
          bestC = c;
          bestGain = gain;
        }
      }
      tot.set(bestC, tot.get(bestC)! + k.get(i)!);
      community.set(i, bestC);
      if (bestC !== ci) {
        moved = true;
        improvedAny = true;
      }
    }
  }

  const remap = new Map<string, number>();
  let nextId = 0;
  const comm = new Map<string, number>();
  for (const i of keys) {
    const c = community.get(i)!;
    if (!remap.has(c)) remap.set(c, nextId++);
    comm.set(i, remap.get(c)!);
  }
  return { comm, improved: improvedAny };
}

/** Newman modularity Q of a partition — exported for tests/validation. */
export function modularity(
  nodeIds: string[],
  adj: Map<string, [string, number][]>,
  community: Map<string, number>,
  resolution = 1,
): number {
  let m2 = 0;
  for (const i of nodeIds) for (const [, w] of adj.get(i) ?? []) m2 += w;
  if (m2 === 0) return 0;
  const deg = new Map<string, number>();
  for (const i of nodeIds) deg.set(i, (adj.get(i) ?? []).reduce((s, [, w]) => s + w, 0));
  let q = 0;
  for (const i of nodeIds) {
    for (const [j, w] of adj.get(i) ?? []) {
      if (community.get(i) === community.get(j)) {
        q += w - (resolution * deg.get(i)! * deg.get(j)!) / m2;
      }
    }
  }
  return q / m2;
}
